### DISCUSSION TITLE
Add open Explorer windows to Explorer's "Quick access" list

### DISCUSSION THREAD

**Discussion by ch3ckmat3** [ramensoftware/windhawk-mods#571] [Mod ideas]
I have been asking for this feature from MS devs though feedback app and though social media, but to no avail.

The idea is to have an item in the Explorer's "Quick access" list with like "Open Windows', and if you click on it, it can list all the currently open explorer windows in the main area. and clicking on any of that window will navigate to that window. This is very helpful when opening and saving files, as the "Quick access" list is also available in the standard File dialogs.

Thoughts?

**Comment by saalkom** [comment]
See https://github.com/NattyNarwhal/OpenWindows

This creates a virtual folder in `This PC`, but you can't pin it to Quick Access (attempting to do so pins `%TEMP%` instead).

**Reply by ch3ckmat3** [reply]
Ah interesting! Thank you.

This code base can be used to implement the mod with better UX.

**Comment by saalkom** [comment]
I think the functionality you are looking for could be attained by making the _`Open Windows`_ virtual folder (CLSID: `{E477F21A-D9F6-4B44-AD43-A95D622D2910}`) provided by NattyNarhwal's _OpenWindows_ shell namespace extension (linked above) a delegate folder of the _`Recent folders`_ (i.e. _`Quick Access`_) virtual folder (CLSID: `{3936e9e4-d92c-4eee-a85a-bc16d5ea0819}`).

Delegate folders are shell folders (can be either virtual or actually extant in the filesystem) whose contents are displayed in a virtual folder as if the latter virtual folder actually directly contained every individual  item in the delegate folder. Items in delegate folders appear in the shell namespace hierarchy as immediate child items of the virtual folder that the delegate folder in question is a delegate folder of. For example, if you navigate to `shell:::{b155bdf8-02f0-451e-9a26-ae317cfd7779}` in Explorer (either by entering that location in Explorer's address bar or the Run dialog, or by executing `explorer.exe shell:::{b155bdf8-02f0-451e-9a26-ae317cfd7779}`, or by creating a new folder (safe to delete afterwards, because this new folder is only a reference to a virtual folder and virtual folders don't actually exist in the filesystem anyway) anywhere named `.{b155bdf8-02f0-451e-9a26-ae317cfd7779}`), you will be taken to a virtual folder whose contents will appear in _`This PC`_. You can't drag and drop to this folder, but by default the current user's _`Network Shortcuts`_ folder (i.e. `%AppData%\Microsoft\Windows\Network Shortcuts` / `%UserProfile%\NetHood`, which does actually exist in the filesystem) is itself actually a delegate folder of the delegate folder that you just navigated to, and so anything in _`Network Shortcuts`_ appears in that delegate folder and consequently also appears in _`This PC`_ (if you have grouping by type enabled in _`This PC`_, you'll see that these items appear under the _`Network locations`_ category).

The contents of delegate folders can also appear in the navigation pane (Microsoft makes use of different delegate folders to dynamically list different kinds of removable media and portable devices in the navigation pane, for example) if the delegate folder's corresponding registry key under `HKCR\CLSID` contains a DWORD `System.IsPinnedToNameSpaceTree` = `1`. (I think the mechanism that makes them appear there is somewhat unique because as far as I'm aware the navigation pane (i.e. the "`NameSpaceTree`") doesn't correspond to any shell folder (either virtual or actually extant in the filesystem) and ultimately only really exists as these scattered DWORD values under various sub-keys of `HKCR\CLSID`). I mention this because I think it would feel counterintuitive if currently-open folders appeared to be permanently pinned to _`Quick Access`_ — what with the little pin icons beside their names and all — and so a more desirable solution may be to show currently-open folders directly in the navigation pane. Currently-open windows won't appear under _`Home`_ if this manner of implementation is chosen, although that could probably be made possible (if desired) by making the _`Open Windows`_ virtual folder a delegate folder of _`Home`_ as well.

I'm not certain how to designate a virtual folder as a delegate folder of either a virtual folder or of the navigation pane, but, given that the default delegate folders are defined in the registry, presumably this can just be achieved via a registry hack if you already have NattyNarhwal's _OpenWindows_ shell namespace extension installed. I may experiment with this myself and I'll happily provide a `.reg` file for your convenience if I'm successful, but I'm not a programmer so I won't be creating a Windhawk mod myself.



Further reading:

- https://winreg-kb.readthedocs.io/en/latest/sources/explorer-keys/Delegate-folders.html
- https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/ns-shobjidl_core-delegateitemid
- https://github.com/libyal/libfwsi/blob/main/documentation/Windows%20Shell%20Item%20format.asciidoc#46-delegate-folder-shell-item

**Comment by Immortal-Sty** [comment]
Maybe you can try [Quicker](https://getquicker.net/) or [Listary](https://www.listary.com/).
